File: pufferlib/ocean/drone_pickplace/drone_pickplace.py
'''Drone Pick and Place Environment for the REVEL x LycheeAI Hackathon

Action Space (10 discrete actions):
0: MOVE_FORWARD - Move forward relative to drone heading
1: MOVE_BACKWARD - Move backward relative to drone heading
2: MOVE_LEFT - Strafe left relative to drone heading
3: MOVE_RIGHT - Strafe right relative to drone heading
4: MOVE_UP - Increase altitude
5: MOVE_DOWN - Decrease altitude
6: ROTATE_LEFT - Yaw counter-clockwise
7: ROTATE_RIGHT - Yaw clockwise
8: GRIPPER_OPEN - Open gripper to release objects
9: GRIPPER_CLOSE - Close gripper to grasp objects

Observation Space (45 continuous values):
- Drone state (14): position(3), velocity(3), quaternion(4), angular_velocity(3), gripper_state(1)
- Object states (21): For each of 3 objects - position(3), velocity(3), status(1)
- Target zones (8): For each of 2 targets - position(3), has_object(1)
- Task info (2): time_remaining(1), task_progress(1)
'''

import gymnasium
import numpy as np

import pufferlib
from pufferlib.ocean.drone_pickplace import binding

class DronePickPlace(pufferlib.PufferEnv):
    def __init__(self, num_envs=1, render_mode=None, log_interval=128, 
                 num_drones=1, num_objects=1, num_targets=1,
                 world_size=2.0, max_height=1.5, max_steps=500,
                 buf=None, seed=0, **kwargs):
        
        # Fixed to 45 observations (must match C code)
        # The C code always outputs 45 observations regardless of object/target count
        obs_per_drone = 45
        total_obs_size = num_drones * obs_per_drone
        
        # Single agent wrapper for now (even with multiple drones)
        # Can be extended to multi-agent later
        self.single_observation_space = gymnasium.spaces.Box(
            low=-np.inf, high=np.inf,
            shape=(total_obs_size,), 
            dtype=np.float32
        )
        
        # 10 discrete actions for quadcopter control
        # Movement: forward, backward, left, right, up, down (6)
        # Rotation: yaw left, yaw right (2)
        # Gripper: open, close (2)
        self.single_action_space = gymnasium.spaces.Discrete(10)
        
        self.render_mode = render_mode
        self.num_agents = num_envs
        self.log_interval = log_interval
        
        # Store environment parameters
        self.num_drones = num_drones
        self.num_objects = num_objects
        self.num_targets = num_targets
        self.world_size = world_size
        self.max_height = max_height
        self.max_steps = max_steps
        
        super().__init__(buf)
        
        # Initialize C environments
        self.c_envs = binding.vec_init(
            self.observations, self.actions, self.rewards,
            self.terminals, self.truncations, num_envs, seed,
            num_drones=num_drones,
            num_objects=num_objects,
            num_targets=num_targets,
            world_size=world_size,
            max_height=max_height,
            max_steps=max_steps
        )
    
    def reset(self, seed=0):
        binding.vec_reset(self.c_envs, seed)
        self.tick = 0
        return self.observations, []
    
    def step(self, actions):
        self.tick += 1
        
        # Handle multi-drone case if needed
        if self.num_drones > 1:
            # For multi-drone, actions should be reshaped appropriately
            # For now, we'll replicate the action for all drones
            expanded_actions = np.repeat(actions, self.num_drones, axis=0)
            self.actions[:] = expanded_actions
        else:
            self.actions[:] = actions
        
        binding.vec_step(self.c_envs)
        
        info = []
        if self.tick % self.log_interval == 0:
            info.append(binding.vec_log(self.c_envs))
        
        return (self.observations, self.rewards,
                self.terminals, self.truncations, info)
    
    def render(self):
        binding.vec_render(self.c_envs, 0)
    
    def close(self):
        binding.vec_close(self.c_envs)

# Multi-agent version for future extension
class DronePickPlaceMultiAgent(DronePickPlace):
    def __init__(self, num_envs=1, render_mode=None, log_interval=128,
                 num_drones=2, num_objects=3, num_targets=2,
                 world_size=2.0, max_height=1.5, max_steps=1000,
                 buf=None, seed=0):
        
        super().__init__(
            num_envs=num_envs * num_drones,  # Each drone is an agent
            render_mode=render_mode,
            log_interval=log_interval,
            num_drones=1,  # Each "env" is now a single drone
            num_objects=num_objects,
            num_targets=num_targets,
            world_size=world_size,
            max_height=max_height,
            max_steps=max_steps,
            buf=buf,
            seed=seed
        )
        
        self.num_agents = num_envs * num_drones


def test_environment():
    """Test the environment with random actions"""
    import time
    
    print("Testing Drone Pick & Place Environment")
    print("=" * 50)
    
    # Single environment test
    env = DronePickPlace(num_envs=1, render_mode="human")
    obs, _ = env.reset()
    
    print(f"Observation shape: {obs.shape}")
    print(f"Action space: {env.single_action_space}")
    print(f"Observation space: {env.single_observation_space}")
    
    steps = 0
    episode_reward = 0
    
    for _ in range(1000):
        # Random action
        action = np.random.randint(0, 10, size=(1,))
        
        obs, reward, done, truncated, info = env.step(action)
        episode_reward += reward[0]
        steps += 1
        
        if render_mode == "human":
            env.render()
            time.sleep(0.01)
        
        if done[0] or truncated[0]:
            print(f"Episode ended. Steps: {steps}, Reward: {episode_reward:.2f}")
            obs, _ = env.reset()
            steps = 0
            episode_reward = 0
    
    env.close()
    
    # Performance test with multiple environments
    print("\nPerformance Test")
    print("-" * 50)
    
    N = 256  # Number of parallel environments
    env = DronePickPlace(num_envs=N)
    env.reset()
    
    CACHE = 1024
    actions = np.random.randint(0, 10, (CACHE, N))
    
    i = 0
    start = time.time()
    total_steps = 0
    
    while time.time() - start < 10:
        env.step(actions[i % CACHE])
        total_steps += N
        i += 1
    
    elapsed = time.time() - start
    sps = int(total_steps / elapsed)
    print(f'DronePickPlace SPS: {sps:,} ({N} environments)')
    print(f'Steps per environment: {total_steps // N}')
    
    env.close()


if __name__ == '__main__':
    test_environment()

File: pufferlib/ocean/drone_pickplace/drone_pickplace.c
/* Main file for testing the drone pick and place environment */

#include "drone_pickplace.h"
#include <stdio.h>
#include <time.h>

int main(int argc, char* argv[]) {
    srand(time(NULL));
    
    printf("Drone Pick & Place Environment Demo\n");
    printf("====================================\n");
    printf("Controls:\n");
    printf("  WASD - Move Forward/Back/Left/Right\n");
    printf("  Q/E - Move Up/Down\n");
    printf("  Z/C - Rotate Left/Right\n");
    printf("  SPACE - Close Gripper\n");
    printf("  R - Open Gripper\n");
    printf("  ESC - Exit\n\n");
    
    // Create environment
    DronePickPlace env = {0};
    env.num_drones = 1;
    env.num_objects = 3;
    env.num_targets = 2;
    env.world_size = 2.0f;
    env.max_height = 1.5f;
    env.max_steps = 1000;
    env.debug_mode = 1;  // Enable debug logging for standalone
    
    // Allocate observation, action, reward, and terminal buffers
    int obs_per_drone = 45;  // Fixed observation size
    env.observations = calloc(env.num_drones * obs_per_drone, sizeof(float));
    env.actions = calloc(env.num_drones, sizeof(int));
    env.rewards = calloc(env.num_drones, sizeof(float));
    env.terminals = calloc(env.num_drones, sizeof(unsigned char));
    
    // Initialize environment
    init(&env);
    c_reset(&env);
    
    // Run simulation
    int total_steps = 0;
    int episodes = 0;
    float total_reward = 0;
    
    while (total_steps < 10000) {
        // Keyboard controls for manual testing
        int action = -1;
        
        if (IsKeyDown(KEY_W)) action = 0;  // MOVE_FORWARD
        else if (IsKeyDown(KEY_S)) action = 1;  // MOVE_BACKWARD
        else if (IsKeyDown(KEY_A)) action = 2;  // MOVE_LEFT
        else if (IsKeyDown(KEY_D)) action = 3;  // MOVE_RIGHT
        else if (IsKeyDown(KEY_Q)) action = 4;  // MOVE_UP
        else if (IsKeyDown(KEY_E)) action = 5;  // MOVE_DOWN
        else if (IsKeyDown(KEY_Z)) action = 6;  // ROTATE_LEFT
        else if (IsKeyDown(KEY_C)) action = 7;  // ROTATE_RIGHT
        else if (IsKeyDown(KEY_SPACE)) action = 9;  // GRIPPER_CLOSE
        else if (IsKeyDown(KEY_R)) action = 8;  // GRIPPER_OPEN
        
        // Use random action if no key pressed
        if (action == -1) {
            action = rand() % 10;
        }
        
        for (int i = 0; i < env.num_drones; i++) {
            env.actions[i] = action;
        }
        
        c_step(&env);
        c_render(&env);
        
        // Accumulate rewards
        for (int i = 0; i < env.num_drones; i++) {
            total_reward += env.rewards[i];
        }
        
        // Check for episode end
        if (env.terminals[0]) {
            episodes++;
            printf("Episode %d completed. Steps: %d, Total Reward: %.2f\n", 
                   episodes, env.current_step, total_reward);
            total_reward = 0;
        }
        
        total_steps++;
        
        // Exit on ESC
        if (IsKeyDown(KEY_ESCAPE)) {
            break;
        }
    }
    
    // Print final statistics
    printf("\n====================================\n");
    printf("Simulation Complete!\n");
    printf("Total Episodes: %d\n", episodes);
    printf("Total Steps: %d\n", total_steps);
    if (env.log.n > 0) {
        printf("Average Performance: %.2f\n", env.log.perf / env.log.n);
        printf("Average Score: %.2f\n", env.log.score / env.log.n);
        printf("Grasp Success Rate: %.2f%%\n", 
               (env.log.grasp_success / (env.log.n * env.num_objects)) * 100);
        printf("Placement Success Rate: %.2f%%\n", 
               (env.log.placement_success / (env.log.n * env.num_objects)) * 100);
    }
    
    // Cleanup
    c_close(&env);
    free(env.observations);
    free(env.actions);
    free(env.rewards);
    free(env.terminals);
    
    return 0;
}